### DISCUSSION TITLE
Docker Container for OpenHAB

### DISCUSSION THREAD

**Discussion by Andre1Becker** [uhppoted/uhppoted#46] [Q&A] [answered]
after a quick websearch, i could not find any docker container of uhppoted... 
It would be really nice and useful if someone could create a containered version - that would be fantastic!
I switched to unraid a couple of days ago and totally forgot about this little gem ;-)
I tried to build one my self but it was way out of my compfy zone...

**Comment by Andre1Becker** [comment]
Sorry - im blind ;-) i just noticed there is already some docker stuff here: https://github.com/uhppoted/uhppoted/tree/master/docker

So i will try to figure out what to do with it - but any help is welcome ;-)

**Comment by uhppoted** [comment]
Hi,

I thought about it once - and then realised that every single person I'm aware of that is using _uhppoted_ has a completely different setup and it didn't seem like it would be practical. And also most people only need a subset of the modules (e.g. the CLI and _uhppoted-mqtt_) so it didn't really make sense to bundle everything up.

The _Dockerfiles_ in the _docker_ folder under the _uhppoted_ umbrella project are mainly for my own use during development but yup, they're a place to start at least - it's not particularly difficult to containerise an individual component. 

If you could let me know which particular component you'd like to containerise though, I'm happy to look into adding a _Dockerfile_ for it :-).

**Reply by Andre1Becker** [reply]
Hey! Well I mostly need what you described exactly, I guess :-)

Since there is no PHP or Python inside the container (openHAB), I just need something to send my open door command to the controller. I did that before turning everything to docker via PHP and the uhppote.php script :-) so I guess mqtt to controller would be nice, or is there some way better/ easier?

Thanks for your time and your great effort!

Cheer's,
Andre

**Reply by YBonline** [reply]
I run uhppoted-mqtt which has a Dockerfile and I built a container for it without issue, but having the prebuilt docker container for the ones you already have a Dockerfile for would be much more convenient then having to build one if it isn't too complicated to do that. I know I'm currently running a few versions behind, mainly because it'd be slightly annoying to update and I've been busy, and there hasn't been any new features I needed, everything is already working perfectly for my current use, but I do generally like to keep software updated

Edit to add - I also remember I was slightly annoyed when I built it that the build process deleted all non-running Docker containers, so I had to rebuild a couple other projects as well because I didn't realize that would happen

**Reply by uhppoted** [reply]
Oops ... you ran `make docker`? That does a `docker prune` because it was intended as just a development thing and I didn't think anybody other than me would use it - I'll move the `prune` to a separate `docker-clean` target in the _Makefile_.

re. prebuilt containers. I'll take another look at it - the only real issue is _uhppoted.conf_ but maybe it's as simple as just providing a default and then it can be customised with an overlay or somesuch. Will think on it - but if you have any suggestions?

**Reply by uhppoted** [reply]
Tracking issue: [Issue #47: Prebuilt Docker containers](https://github.com/uhppoted/uhppoted/issues/47)

**Reply by Andre1Becker** [reply]
Well, I'm very new to the docker-game but why not keep the config files persistent via "mount" or "volume" option in docker? I thought I've read about that somewhere in the docker docs. Many other images are doing something similar (like nextcloud, openHAB, pi-hole and so on).

I'm using unRAID as my base system btw...

**Reply by uhppoted** [reply]
_mount_ is one option - although I was trying to use it recently and it was harder than it should have been, at least on a Mac.

re. unRAID. Hadn't come across it until you mentioned it - looks very slick!

**Reply by YBonline** [reply]
Yeah, typically volumes is how this is done on almost every project I'm running. As a piece of advise, make sure you assign the entire folder to the volume. For uhppoted, I believe its 2 storage directories are:
- /etc/uhppoted
- /var/lib/uhppoted

If using docker-compose (which I find easier to make it to distribute, you can post a zip file with a sample uhppoted.conf, a docker-compose such as the sample below, and then a user can just edit the config file, type `docker-compose up` and then they'll have a working instance up within seconds, this would be a sample docker-compose.yml:
```
version: "3.9"
services:
  uhppoted-mqtt:
    container_name: uhppoted-mqtt
    privileged: true # this may not be necessary for all setups
    restart: unless-stopped
    image: ghcr.io/uhppoted/uhppoted-mqtt:latest
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /etc/uhppoted:/etc/uhppoted
      - /var/lib/uhppoted:/var/lib/uhppoted
    ports:
      - "5000:5000"
```

For volumes, the first half is the location on the local HDD, after the `:` is the location inside the container.

ghcr is github container repository, allowing you to keep everything within github for your development as well
https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry

**Reply by uhppoted** [reply]
Cool, thanks YB - that's really helpful!

**Comment by uhppoted** [comment]
One of these days I'll think about getting around to doing an _OpenHAB_ integration but that ecosystem is complex so it's a long way down the road.

For now though, setting up a whole MQTT broker and client for an _open-door_ command seems a bit like overkill - there are a couple of other more lightweight approaches that might be worth looking into first:

- [uhppoted-nodejs](https://github.com/uhppoted/uhppoted-nodejs) with the [Javascript add-on](https://www.openhab.org/addons/automation/jsscripting)
- [uhppoted-rest](https://github.com/uhppoted/uhppoted-rest) running in a Docker container (there's already a recipe for that at least) with an [HTTP action](https://www.openhab.org/docs/configuration/actions.html#http-actions)
- [uhppote-cli](https://githhub.com/uhppoted/uhppote-cli) with an [_Exec_ action](https://www.openhab.org/docs/configuration/actions.html#exec-actions)

If it were me I would probably start with the _Exec_ action and the CLI - if that works it's probably the least complicated way of doing anything. If your requirements are more complex then the [JS Scripting](https://www.openhab.org/docs/configuration/jsr223.html#overview) looks promising - the documentation says it can import _NodeJS_ modules which is a good start.

**Reply by Andre1Becker** [reply]
Perfect!
I will look into that. And if I'm not to lazy :-) ill write a little tutorial here :-)

**Reply by uhppoted** [reply]
A tutorial would be **awesome** :-)

**Reply by Andre1Becker** [reply, marked as answer]
So here is a ~~short~~ documentation of steps i did to get it running with Docker, MQTT and openHAB on unRAID ;-)

# uhppoted and Docker

1. Download [uhppoted-mqtt Repository](https://github.com/uhppoted/uhppoted-mqtt/releases) and extract the `uhppoted-mqtt` file from the linux folder
2. Download [the MQTT Docker Folder](https://github.com/uhppoted/uhppoted-mqtt/tree/main/docker)
3. Extract all to one Folder, so it that looks like this:
![image](https://github.com/uhppoted/uhppoted/assets/25841903/227d41d3-b021-4d6c-87b8-a508deeafbfb)
4. Modify the `uhppoted.conf` for your configuration, including adding your controllers under the `# Devices` section (optional but recommended):

```
# DEVICES
**fyi: 123456789 is the Serial number of your controller**
UT0311-L0x.123456789.address = 172.16.1.102:60000
UT0311-L0x.123456789.door.1 = Frontdoor
UT0311-L0x.123456789.door.2 = Garage
```
5. Modify your _Dockerfile_ to look like this:

```
FROM alpine

RUN mkdir -p /etc/uhppoted
RUN mkdir -p /opt/uhppoted
RUN mkdir -p /etc/uhppoted/mqtt
RUN mkdir -p /etc/uhppoted/mqtt/rsa
RUN mkdir -p /etc/uhppoted/mqtt/rsa/signing
RUN mkdir -p /etc/uhppoted/mqtt/rsa/encryption
RUN mkdir -p /var/uhppoted

COPY uhppoted.conf                   /etc/uhppoted
COPY hivemq.pem                      /etc/uhppoted
COPY secure/client.key               /etc/uhppoted/mqtt
COPY secure/client.cert              /etc/uhppoted/mqtt
COPY secure/rsa/signing/mqttd.key    /etc/uhppoted/mqtt/rsa/signing
COPY secure/rsa/encryption/mqttd.key /etc/uhppoted/mqtt/rsa/encryption
COPY uhppoted-mqtt /opt/uhppoted
WORKDIR /opt/uhppoted 

VOLUME /etc/uhppoted /var/uhppoted

ENTRYPOINT /opt/uhppoted/uhppoted-mqtt --console
```

6. Install docker if not already done.
7. Open a terminal or CMD and change into your folder where the files stored and
8. Run `docker build -t uhppoted-mqtt .`. This will create a docker image.
9. Tag the image `docker tag uhppoted-mqtt YourDockerUsername/uhppoted-mqtt:latest`. This tag is used to upload your image to docker hub.
10. Push the image to your docker hub repo with: `docker image push YourDockerUsername/uhppoted-mqtt:latest`
 
# unRAID

1. Login to your unRAID WEB GUI
2. Now login to docker from unRAID Terminal. to do so: 
3. Click the Button for unRAID Terminal in the unRAID UI 

![image](https://github.com/uhppoted/uhppoted/assets/25841903/47af79f0-aa90-4ceb-a3bf-a8b9d0ff9b81)

4. In the shell window, Type "docker login" and press enter. You will be prompted for your name and password.

![image](https://github.com/uhppoted/uhppoted/assets/25841903/4e790694-d8bc-4561-b545-b52af85d686b)

**NB**: Please pay attention to the **WARNING** - see the [docker login](https://docs.docker.com/engine/reference/commandline/login/#credentials-store) documentation for a more secure approach or use `docker logout` once finished with your setup.

Now you can close or logout from the unRAID terminal.

5. Next, create new container, therefore go to the Docker tab and press "Add Container"

![image](https://github.com/uhppoted/uhppoted/assets/25841903/c6ef2aab-bc84-4c62-bcf3-b81e2094e7cc)

6. In the "Repository:" field enter your repository in the format `docker.io/[DockerUsername]/[DockerRepositoryname]`

![image](https://github.com/uhppoted/uhppoted/assets/25841903/53c68e23-064e-451f-baa9-4314897bc259)

7. Configure your container as needed. e.g. add a volume mount so you can modify the `uhppoted.conf` without republishing the image:

![image](https://github.com/uhppoted/uhppoted/assets/25841903/55309f32-75d4-4f3c-b354-111d0ad9c678)

Also, create a file named `cards` under `/etc/uhppoted/mqtt` and put the `card-id` you want to use in there. In my example its `100100100.*`.

If you would like a custom logo for your container, click the toggle switch in the upper right and switch to `advanced view`:

![image](https://github.com/uhppoted/uhppoted/assets/25841903/68155c87-8399-47a0-b429-5e794330e18c)

and put this URL: `https://avatars.githubusercontent.com/u/61560842?v=4` into the `icon url` field:

![image](https://github.com/uhppoted/uhppoted/assets/25841903/bbf8eec8-4706-4e12-9167-df6f851360ac)

8. Then Press "DONE". If everything checks out, you should find a running container in your Docker tab:

![image](https://github.com/uhppoted/uhppoted/assets/25841903/77a0c8d5-8de4-4271-a071-175b6d73fb72)

If not, left click the container logo to open the log output for troubleshooting:

![image](https://github.com/uhppoted/uhppoted/assets/25841903/39c39924-12e5-455d-84f3-dc60ab78af63)

# openHAB

Now to the openHAB side of things. This should be very easy and straight forward :-)
I assume that you already have setup a MQTT bridge in openHAB and also have a MQTT Broker running. If not, check out [openHAB MQTT binding](https://www.openhab.org/addons/bindings/mqtt/) documentation to get you started.

1. Create a new Thing with mqtt-binding
Click on Things, then on the Plus Icon in the bottem right corner:

![image](https://github.com/uhppoted/uhppoted/assets/25841903/f743f98e-5034-4b46-9b43-e2cd5996841f)

2. Click on `mqtt-binding`

![image](https://github.com/uhppoted/uhppoted/assets/25841903/ab9dbacc-d190-496e-952c-304bafc11d88)

3. Choose `generic mqtt Thing`

![image](https://github.com/uhppoted/uhppoted/assets/25841903/29c831c7-8424-494a-a9f5-815863e5e82a)

4. Give it a name and click `create Thing`:

![image](https://github.com/uhppoted/uhppoted/assets/25841903/c3ecba23-ced4-42d5-be17-5bc3aa65d5b9)

5. Create a new Channel with channel type `on/off switch`

![image](https://github.com/uhppoted/uhppoted/assets/25841903/bdf05b11-89fd-47eb-9d4a-23cde2c50535)

![image](https://github.com/uhppoted/uhppoted/assets/25841903/4c7a4af3-5877-4f6c-afe8-12924ea6b855)

![image](https://github.com/uhppoted/uhppoted/assets/25841903/00d42d8e-ef5a-44ee-ae59-31fdd959b44c)


6. Add a command topic and a message as open/on value:
 
![image](https://github.com/uhppoted/uhppoted/assets/25841903/62254dab-9712-45d4-b631-59ffe01d2143)

Here is a example message for opening a door: 

```
{ "message": { "request": { "client-id": "openHABuhppoted", "request-id": "AH173635G3", "reply-to": "uhppoted/reply/97531", "device-id": 123456789, "door": 2, "card-number": 100100100 } } }
 ```

_**Note**_:
`client-id`, `request-id` and `reply-to topic number` can be anything you like.
most important parts are the `device-id` which is the `serial-number` of the controller, `door id` and the `card-number`.

Commands for additional functionality are documented in the [uhppoted-mqtt documentation](https://github.com/uhppoted/uhppoted-mqtt/tree/main/documentation/commands) .

Now you just need to link a switch item to the channel and boom - open sesame ;-).

# Troubleshooting Tips:
- Look in the Logfile mentioned above
- Use a Tool like [MQTT-Explorer](https://mqtt-explorer.com/) or [MQTTX](https://mqttx.app/) to see whats going on mqtt wise ;-)

**Comment by uhppoted** [comment]
You make it sound almost easy! Thanks *so* much Andre .. really appreciated! It's so nice to have something to point to for the occasional OpenHAB question I get.

I'll add a note to the README with a link to this so that people don't have to hunt for it.

**Reply by Andre1Becker** [reply]
Your welcome :-) the least thing I could do to give something back. 

I will polish the documentation a bit more, when there is time ;-)

**Reply by uhppoted** [reply]
Documentation is the **worst** :-) .. it takes **forever**!! But even just knowing that something is possible and having a place to start helps a huge amount.

**Reply by Andre1Becker** [reply]
Well i think im done now - but is it okay to read (and hopefully understandable)? Feedback is welcome :-)

**Reply by uhppoted** [reply]
I think it looks fine Andre - I went through it and fixed a couple of small typos and minor grammatical things, but it seems perfectly readable and clear to me! 

And it's a very useful piece of documentation - thank you very much for putting in the time and effort!
